### ISSUE TITLE
Extend `use_self` to structs

### ISSUE THREAD

**Issue by nickdrozd** [rust-lang/rust-clippy#15555]
### Summary

Existing lint `use_self` check `impl` blocks for places where `Self` could be used instead of explicit repetition of type.

The same can be done with recursive structs. For example:

```rust
struct X {
    x: Option<Box<X>>,  // <-- repetition
}
```

```rust
struct X {
    x: Option<Box<Self>>,  // <-- no repetition
}
```

`Self` is very nice to have if the struct is modified to be generic.

```rust
struct X<'q, T: From<usize>> {
    t: &'q T,
    x: Option<Box<X<'q, T>>>,  // <-- yuck
}
```

Opened as false negative, but might also be a "feature-request in a trench coat" 😆

### Lint Name

use_self

### Reproducer

Should raise `use_self`:

```rust
struct X {
    x: Option<Box<X>>,  // <-- Self
}
```

```rust
struct X<'q, T: From<usize>> {
    t: &'q T,
    x: Option<Box<X<'q, T>>>,  // <-- Self
}
```

### Version

```text
rustc 1.91.0-nightly (6ba0ce409 2025-08-21)
binary: rustc
commit-hash: 6ba0ce40941eee1ca02e9ba49c791ada5158747a
commit-date: 2025-08-21
host: x86_64-unknown-linux-gnu
release: 1.91.0-nightly
LLVM version: 21.1.0
```
